src/Core/StaticConfigTrait.php

```diff
@@ -249,23 +249,53 @@ public static function parseDsn($dsn)
             throw new InvalidArgumentException('Only strings can be passed to parseDsn');
         }
 
-        $pattern = '/^(?P<scheme>[\w\\\\]+):\/\/((?P<username>.*?)(:(?P<password>.*?))?@)?' .
-            '((?P<host>[^?#\/:@]+)(:(?P<port>\d+))?)?' .
-            '(?P<path>\/[^?#]*)?(\?(?P<query>[^#]*))?(#(?P<fragment>.*))?$/';
-        preg_match($pattern, $dsn, $parsed);
-
-        if (!$parsed) {
-            throw new InvalidArgumentException("The DSN string '{$dsn}' could not be parsed.");
+        $scheme = $noScheme = '';
+        if (preg_match('{^(?<scheme>[\w\\\]+)(:(//)?)(?P<noScheme>(.*))}', $dsn, $matches)) {
+            $scheme = $matches['scheme'];
+            $noScheme = $matches['noScheme'];
         }
-        foreach ($parsed as $k => $v) {
-            if (is_int($k)) {
-                unset($parsed[$k]);
+
+        $parsed = parse_url('file://' . $noScheme);
+        if ($parsed === false) {
+            // Failed to parse.
+            // Try to encode the userinfo if the DSN string contains '@' marks.
+            for ($p = 0; ($p = strpos($noScheme, '@', $p)) !== false; ++$p) { // @codingStandardsIgnoreLine
```

Comment:
But wouldn't `parse_url()` cover most use cases/scenarios has it historically has? The regex would only be used if the user/password contained `#` or `?` which cause `parse_url()` to behave incorrectly.